safekeeper/src/send_wal.rs

```diff
@@ -589,14 +591,22 @@ impl<IO: AsyncRead + AsyncWrite + Unpin> WalSender<'_, IO> {
     async fn wait_wal(&mut self) -> Result<(), CopyStreamHandlerEnd> {
         loop {
             self.end_pos = self.end_watch.get();
-            if self.end_pos > self.start_pos {
-                // We have something to send.
+            let have_something_to_send = (|| {
```

Comment:
I don't want to do `continue` because I want to send keepalives

Reply:
I wanted to have the connection be successful, but just have the safekeeper not sending data (but still send keepalives). I think just bailing the from START_REPLICATION will terminate the connection.

Reply:
> I think just bailing the from START_REPLICATION will terminate the connection.

What are the downsides (in terms of testing) of this compared to successful connections that send no data but send keepalives?

Reply:
I think they test slightly different things. One exercises connection retry logic and the other exercises polling logic. Though maybe for this bug it would still reproduce it

Reply:
Ok, looked at failpoints api, haven't found an easy way to do it simpler. For now I'm giving up on this, but would be nice to revisit this testing later